README.md
# deezer-control
Google Chrome extension to control music playback on Deezer.com


CHANGELOG
2.16 - 

2.15 - 04/02/2020
 * Added support of the new Chrome Global Media Controls pane

2.14 - 28/10/2018
 * Use Deezer events instead of a DOM observer to monitor track changes

2.13 - 01/07/2018
 * Fixed issue with favorite icon not updating.

2.12 - 28/11/2016
 * Fixed issue that prevented extension from tracking the play / pause button
 
2.11 - 07/11/2016
 * Fixed issue that prevented extension from reading the track title

2.10 - 07/11/2016
 * Fixed issue that prevented extension from loading

2.9 - 02/09/2016
 * Fixed CSS for large theme that prevented letters from being fully displayed 
 
 2.8 - 02/09/2016
 * Removed accidental logging of MutationRecords 
 
2.7 - 01/09/2016
 * Removed obsolete notifications option "never hides" (obsolete, because the notification center no longer exists)
 * Updated options style to match new Deezer CSS
 * Fixed broken info update on track change 

2.6 - 06/11/2015
 * Improvement: love icon is now correctly updated in real time

2.5 - 21 Jan 2015
 * Fix: updated extension to match latest Deezer API
 
2.4 - 18 Oct 2014 
 * Fix: new Deezer is now properly supported
 * Improvement: liking a song is now possible
 * Improvement: added support of radiofg.com
 
2.3 - 04 Oct 2014 
 * Improvement: added Italian and Portuguese (Brazil) translations (thanks to Alexandre Vicenzi)
 * Improvement: added support of new Deezer 
 * Improvement: added support of SoundCloud.com

2.2 - 30 Jul 2014
 * Fix: https://bitbucket.org/metoule/deezer_control/issue/3/deezer-control-causing-issues-with-chrome
 * Fix: no longer try to execute content script on chrome:// URLs
 * Improvement: multiple deezer tabs can be opened at the same time
 * Improvement: no longer rely on URLs to detect if another player is playing
 * Improvement: reloading the extension no longer makes the extension fail

2.1 - 09 Jun 2014
 * Follow Deezer JS change

2.0 - 02 Jun 2014
 * New, flatter, icons
 * Fix: prevent hotkeys from being seen twice
 * Fix: orange.deezer.com users can now use the extension as expected
 * Improvement: hotkeys can be used even if the page isn't fully loaded
 * Improvement: hotkeys can now be disabled
 * Improvement: use less memory
 * Style: keep in sync with Deezer style
 * Style: buttons are now more consistent across the extension
 * Cleanup: old notification system completely removed
 
1.8 - 
 * Extension keeps working after update (no longer requires a page refresh)
 * Limit the number of useless notification refresh 
 * Listener on player changes is now in its own file
 * Re-added notifications display options  

1.7 - 24 Sep 2013 
 * Improved first time install:
 * Deezer tab no longer needs to be refreshed for the popup to display content 
 * When hotkeys are activated for the first time, hotkeys.js is injected on all opened pages 
 
1.6 - 13 Sep 2013
 * New notifications system to support latest Chrome versions
 * Use latest Chrome API
 * Player listener rewritten to fire less events 
 
1.5 - 08 Aug 2013
 * Update listener script to match Deezer update

1.3/1.4 - 30 May 2013
 * Correctly extend the show function to avoid breaking the jQuery chaining

1.2 - 
 * Song and artist names are now links to their respective pages
 * It's now possible to skin the notifications
 * Don't preload notifications covers if not needed
 * Correctly monitor play / pause changes

1.1 - 27 Jan 2013
 * Fixed https://bitbucket.org/metoule/deezer_control/issue/1/package-incorrect

1.0 - 19 Nov 2012
 * Replaced badge text from NEW to version to work around bug https://code.google.com/p/chromium/issues/detail?id=154270 
 * Rewrote options' page explanation text
 * Improved Deezer tab mechanism to allow reuse
 * Bug fixes for notif always on: close notif if no Deezer tab opened / re-open notif if closed
 * Added jump to Deezer feature

0.1.5 - 05 Oct 2012
 * Fixed typo in french locale

0.1.4 - 28 Sep 2012 
 * Updated style to latest Deezer one
 * Improved the way deezer tabs are counted to avoid the empty popup symptom

0.1.3 - 21 Jul 2012
 * Bubble mouse event on notif mouse over to correctly show cursor on buttons

0.1.2 - 19 Jul 2012 
 * Stop notification timer on mouse over

0.1.1 - 19 Jul 2012
 * Added hot keys support
 * Improved local storage for easier default values
 * Improved notification system to be displayable on hot keys only 
 * Options.html migrated to jQuery
 * Pre cache album covers for faster display
 * Fixed flickering on popup load if change of CSS needed

0.0.8 - 03 Mar 2012 
 * Added script to ease delivery

0.0.7 - 03 Mar 2012
 * Use data-i18n-content in HTML tags to be HTML5 compliant
 * Improved popup display time by using a workaround for bug http://crbug.com/24467
 * Changed default image to an empty CD
 * Change event to listen to in player_listener, from unload to beforeunload, so that we're guaranteed to be called

0.0.6 - 19 Feb 2012
 * Improved notifications mechanism

0.0.5 - 18 Feb 2012
 * Solve issue with erroneous opened tabs count on page refresh
 * Hide next / prev buttons if inactive on Deezer 

0.0.4 - 16 Feb 2012 
 * Added text to welcome page
 * Added notifications
 * Update cover size on popup style change
 * If no deezer tab is opened, a click on browser action opens one

0.0.3 - 07 Feb 2012
 * Set large popup style as default for new installs

0.0.2 - 05 Feb 2012

 * Initial revision 


scripts/options.js

window.addEventListener('hashchange', function(/*e*/) { "use strict"; showSection(location.hash.replace(/^\#/,'')); });
window.addEventListener('load', preparePage);


function sendOptionsChangedMessage()
{
	chrome.runtime.sendMessage({ type: "optionsChanged" }); 
}

function showSaveStatus(statusId)
{
	$("#" + statusId).stop(true, true).text(chrome.i18n.getMessage("options_page_options_saved")).show().fadeOut(1500);
}

function preparePage()
{
	"use strict";
	i18n.process(document);
	
	preparePage_welcome();
	preparePage_style();
	preparePage_hotKeys();
	preparePage_notifs();
	preparePage_misc();

	resetSections();
}


function preparePage_welcome()
{
	"use strict";
	$("#button_rate_extension").attr('href', "https://chrome.google.com/webstore/detail/" + chrome.i18n.getMessage("@@extension_id"));
}

function changeStyle(iPopupStyle)
{
	"use strict";
	loadStyle(iPopupStyle);
}

function preparePage_style()
{
	"use strict";
	
	// create interactivity
	$("#popup_style_chooser").change(function () { changeStyle($("#popup_style_chooser").val()); });
	$("#button_save_style").click(savePopupStyle);

	// restore value 
	$("#popup_style_chooser").val(LOCSTO.popupStyle);
}


function preparePage_hotKeys()
{
	"use strict";
	
	// create interactivity
	$("#button_activate_hotkeys").click(activateHotKeys);
	$("#button_disable_hotkeys").click(disableHotKeys);
	$("#button_save_hotkeys").click(saveHotKeys);

	// restore value
	restoreHotkey('prevHotKey');
	restoreHotkey('playPauseHotKey');
	restoreHotkey('nextHotKey');
	restoreHotkey('addToFavoriteHotKey');
	restoreHotkey('whatZatSongHotKey');
	restoreHotkey('jumpToDeezerHotKey');

	// hot keys are activated only if we have permission on all tabs
	// if we don't permission, show an explanation
	chrome.permissions.contains({ origins: ['<all_urls>'] }, function(granted)
	{
		displayHotKeysOptions(granted);
	});
}


function preparePage_notifs()
{
	"use strict";

	// create interactivity
	$("#button_save_notifications").click(saveNotifications);

	// restore value 
	if (LOCSTO.notifications.onSongChange)
	{
		$('input:radio[name="notifs_show_when"]').filter('[value="on_song_change"]').prop('checked', true);
	}
	else if (LOCSTO.notifications.onHotKeyOnly)
	{
		$('input:radio[name="notifs_show_when"]').filter('[value="on_hotkey_only"]').prop('checked', true);
	}
	else
	{
		$('input:radio[name="notifs_show_when"]').filter('[value="never"]').prop('checked', true);
	}
}


function preparePage_misc()
{
	"use strict";
	
	// create interactivity
	$('#miscLimitDeezerToOneTab > .yes_no_bar').children('button').click(function() 
	{
		if (!$(this).hasClass('btn_selected')) 
		{
			$(this).parent().children('button').toggleClass('btn_selected btn_unselected');
		}
	});
	$("#button_save_misc").click(saveMiscOptions);

	// restore value 
	$("#miscLimitDeezerToOneTab > .yes_no_bar").children('button:eq(0)').toggleClass('btn_unselected', !LOCSTO.miscOptions.limitDeezerToOneTab).toggleClass('btn_selected',  LOCSTO.miscOptions.limitDeezerToOneTab);
	$("#miscLimitDeezerToOneTab > .yes_no_bar").children('button:eq(1)').toggleClass('btn_unselected',  LOCSTO.miscOptions.limitDeezerToOneTab).toggleClass('btn_selected', !LOCSTO.miscOptions.limitDeezerToOneTab);
}


function displayHotKeysOptions(granted)
{
	"use strict";

	$("#hotkey_permission_ko").toggle(!granted);
	$("#hotkey_permission_ok").toggle(granted);
}


function restoreHotkey(iHotKeyName)
{
	"use strict";
	
	var aHotKeySelect = $("#" + iHotKeyName);

	// on click, all button are selected
	aHotKeySelect.children('button').click(function() { $(this).toggleClass('btn_selected btn_unselected'); });
	aHotKeySelect.children('span').css('visibility', 'visible');

	// last button: click reveals the textbox to enter new key
	aHotKeySelect.children('button:eq(3)').unbind('click').click(function()
	{
		$(this).parent().children('span').text("");
		$(this).css('display', 'none');
		$(this).parent().children('input:eq(0)').css('display', 'inline').focus();
	});

	// on key up in the input box, update last button with the new key code
	aHotKeySelect.children('input:eq(0)').keydown(function()
	{
		// if ctrl, shift or alt, prevent update and warn user
		if (event.keyCode >= 16 && event.keyCode <= 18)
		{
			$(this).parent().children('span').text(chrome.i18n.getMessage("options_page_tab_hotkeys_warning_illegal_hotkey"));
		}
		else
		{
			$(this).parent().children('button:eq(3)').text(convertKeyCode(event.keyCode));
			$(this).next().val(event.keyCode); // input:eq(1)
		}

		// redisplay
		$(this).parent().children('button:eq(3)').css('display', 'inline').focus().blur();
		$(this).css('display', 'none').val("");
	});

	aHotKeySelect.children('input:eq(0)').blur(function()
	{
		$(this).css('display', 'none').val("");
		$(this).parent().children('button:eq(3)').css('display', 'inline').focus().blur();
	});


	// create display
	aHotKeySelect.children('button:eq(0)').toggleClass('btn_unselected', !LOCSTO[iHotKeyName].shiftKey).toggleClass('btn_selected', LOCSTO[iHotKeyName].shiftKey);
	aHotKeySelect.children('button:eq(1)').toggleClass('btn_unselected', !LOCSTO[iHotKeyName].ctrlKey).toggleClass('btn_selected', LOCSTO[iHotKeyName].ctrlKey);
	aHotKeySelect.children('button:eq(2)').toggleClass('btn_unselected', !LOCSTO[iHotKeyName].altKey).toggleClass('btn_selected', LOCSTO[iHotKeyName].altKey);
	aHotKeySelect.children('button:eq(3)').text(convertKeyCode(LOCSTO[iHotKeyName].keyCode));

	aHotKeySelect.children('input').attr('size', '3').css('display', 'none');
	aHotKeySelect.children('input:eq(1)').val(LOCSTO[iHotKeyName].keyCode);
}


function savePopupStyle()
{
	"use strict";
	
	LOCSTO.popupStyle = $("#popup_style_chooser").val();
	LOCSTO.savePopupStyle();
	sendOptionsChangedMessage();
	
	// Update status to let user know options were saved.
	showSaveStatus("status_style");
}


function saveHotKeys()
{
	"use strict";
	
	storeHotKey('prevHotKey');
	storeHotKey('playPauseHotKey');
	storeHotKey('nextHotKey');
	storeHotKey('addToFavoriteHotKey');
	storeHotKey('whatZatSongHotKey');
	storeHotKey('jumpToDeezerHotKey');
	LOCSTO.saveHotKeys();
	sendOptionsChangedMessage();

	// Update status to let user know options were saved.
	showSaveStatus("status_hotkeys");
}


function storeHotKey(iHotKeyName)
{
	"use strict";
	
	var aHotKeyDiv = $("#" + iHotKeyName);
	LOCSTO[iHotKeyName].shiftKey = aHotKeyDiv.children('button:eq(0)').hasClass('btn_selected');
	LOCSTO[iHotKeyName].ctrlKey  = aHotKeyDiv.children('button:eq(1)').hasClass('btn_selected');
	LOCSTO[iHotKeyName].altKey   = aHotKeyDiv.children('button:eq(2)').hasClass('btn_selected');
	LOCSTO[iHotKeyName].keyCode  = parseInt(aHotKeyDiv.children('input:eq(1)').val(), 10);
}


function saveHotkeysPermission(granted)
{
	LOCSTO.miscOptions.hasHotkeysPermission = granted;
	LOCSTO.saveMiscOptions();
	sendOptionsChangedMessage();
}


function saveNotifications()
{
	"use strict";
	
	var aNotifsShowWhen = $('input:radio[name="notifs_show_when"]:checked').val();
	
	if (aNotifsShowWhen === 'on_song_change')
	{
		LOCSTO.notifications = { never: false, onSongChange: true, onHotKeyOnly: false };
	}
	else if (aNotifsShowWhen === 'on_hotkey_only')
	{
		LOCSTO.notifications = { never: false, onSongChange: false, onHotKeyOnly: true };
	}
	
	LOCSTO.saveNotifications();
	sendOptionsChangedMessage();

	// Update status_style to let user know options were saved.
	showSaveStatus("status_notifs");

	// reshow notifs so that the user sees the change straight away
	chrome.runtime.sendMessage({ type: "showNotif" });
}


function saveMiscOptions()
{
	"use strict";
	
	// limit deezer to one tab
	LOCSTO.miscOptions.limitDeezerToOneTab = $("#miscLimitDeezerToOneTab > .yes_no_bar").children('button:eq(0)').hasClass('btn_selected');
	LOCSTO.saveMiscOptions();
	sendOptionsChangedMessage();

	// Update status to let user know options were saved.
	showSaveStatus("status_misc");
}


// deal with chrome permissions
function activateHotKeys()
{
	"use strict";
	chrome.permissions.request({ origins: [ '<all_urls>' ] }, function(granted)
	{
		if (granted)
		{
			chrome.runtime.sendMessage({ type: "injectHotKeysJsOnAllTabs" }); 
		}
		displayHotKeysOptions(granted);
		saveHotkeysPermission(granted);
	});
}


function disableHotKeys()
{
	"use strict";
	chrome.permissions.remove({ origins: [ '<all_urls>' ] }, function(removed)
	{
		displayHotKeysOptions(!removed);
		saveHotkeysPermission(!removed);
	});
}

// navigation bar
function resetSections()
{
	"use strict";
	
	$('#tab_chooser > nav > a:first').addClass('currentone');
	$('#center_block > section:first').css('display', 'block');
	if(location.hash)
	{
		showSection(location.hash.replace(/^\#/, ''));
	}
}


function showSection(id)
{
	"use strict";
	var aNewLeft;
	
	$('#tab_chooser > nav > a').removeClass('currentone');
	$('#center_block > section:not(#' + id + ')').css('display', 'none');
	$('#' + id).css('display', 'block');
	$('#' + id + '_nav').addClass('currentone');

	aNewLeft = $('#' + id + '_nav').position().left - $('#tab_chooser > nav > a:first').position().left + 73;
	$('#arrow_line > span').css('left', aNewLeft + 'px');
}


function convertKeyCode(iKeyCode)
{
	"use strict";
	var aMapKeyToText = {
		   8: "keycode_text_backspace",
		   9: "keycode_text_tab",
		  13: "keycode_text_enter",
		  16: "keycode_text_shift",
		  17: "keycode_text_ctrl",
		  18: "keycode_text_alt",
		  19: "keycode_text_pause_break",
		  20: "keycode_text_caps_lock",
		  27: "keycode_text_escape",
		  32: "keycode_text_space",
		  33: "keycode_text_page_up",
		  34: "keycode_text_page_down",
		  35: "keycode_text_end",
		  36: "keycode_text_home",
		  37: "keycode_text_left_arrow",
		  38: "keycode_text_up_arrow",
		  39: "keycode_text_right_arrow",
		  40: "keycode_text_down_arrow",
		  42: "keycode_text_print_screen",
		  45: "keycode_text_insert",
		  46: "keycode_text_delete",
		  91: "keycode_text_left_window_key",
		  92: "keycode_text_right_window_key",
		  93: "keycode_text_select_key",
		  96: "keycode_text_numpad_0",
		  97: "keycode_text_numpad_1",
		  98: "keycode_text_numpad_2",
		  99: "keycode_text_numpad_3",
		 100: "keycode_text_numpad_4",
		 101: "keycode_text_numpad_5",
		 102: "keycode_text_numpad_6",	
		 103: "keycode_text_numpad_7",	
		 104: "keycode_text_numpad_8",	
		 105: "keycode_text_numpad_9",	
		 106: "keycode_text_numpad_multiply",	
		 107: "keycode_text_numpad_plus",	
		 109: "keycode_text_numpad_minus",	
		 110: "keycode_text_numpad_dot",	
		 111: "keycode_text_numpad_slash",	
		 112: "keycode_text_F1",	
		 113: "keycode_text_F2",		
		 114: "keycode_text_F3",		
		 115: "keycode_text_F4",		
		 116: "keycode_text_F5",		
		 117: "keycode_text_F6",		
		 118: "keycode_text_F7",		
		 119: "keycode_text_F8",		
		 120: "keycode_text_F9",		
		 121: "keycode_text_F10",		
		 122: "keycode_text_F11",		
		 123: "keycode_text_F12",		
		 144: "keycode_text_num_lock",	
		 145: "keycode_text_scroll_lock",
		 176: "keycode_text_media_key_next",
		 177: "keycode_text_media_key_prev",
		 178: "keycode_text_media_key_stop",
		 179: "keycode_text_media_key_play",
		 186: "keycode_text_semi_column",
		 187: "keycode_text_equal",
		 188: "keycode_text_coma",
		 189: "keycode_text_minus",
		 190: "keycode_text_dot",
		 191: "keycode_text_slash",
		 192: "keycode_text_grave",
		 219: "keycode_text_left_bracket",
		 220: "keycode_text_back_slash",
		 221: "keycode_text_right_bracket",
		 222: "keycode_text_quote"
	};
	
	if (aMapKeyToText.hasOwnProperty(iKeyCode))
	{
		return chrome.i18n.getMessage(aMapKeyToText[iKeyCode]);
	}
	
	return String.fromCharCode(iKeyCode);
}
